operator/validator/controller.go

```diff
@@ -371,55 +374,81 @@ var nonCommitteeValidatorTTLs = map[spectypes.RunnerRole]phase0.Slot{
-			share := c.sharesStorage.Get(nil, msg.GetID().GetDutyExecutorID())
-			if share == nil {
-				return errors.Errorf("could not find validator [%s]", hex.EncodeToString(msg.GetID().GetDutyExecutorID()))
-			}
+	var ncv *committeeObserver
+	item := c.getNonCommitteeValidators(msg.GetID())
+	if item == nil {
+		committeeObserverOptions := validator.CommitteeObserverOptions{
+			Logger:            c.logger,
+			NetworkConfig:     c.networkConfig,
+			ValidatorStore:    c.validatorStore,
+			Network:           c.validatorOptions.Network,
+			Storage:           c.validatorOptions.Storage,
+			FullNode:          c.validatorOptions.FullNode,
+			Operator:          c.validatorOptions.Operator,
+			NewDecidedHandler: c.validatorOptions.NewDecidedHandler,
+		}
+		ncv = &committeeObserver{
+			CommitteeObserver: validator.NewCommitteeObserver(convert.MessageID(msg.MsgID), committeeObserverOptions),
+		}
+		ttlSlots := nonCommitteeValidatorTTLs[msg.MsgID.GetRoleType()]
+		c.committeesObservers.Set(
+			msg.GetID(),
+			ncv,
+			time.Duration(ttlSlots)*c.beacon.GetBeaconNetwork().SlotDurationSec(),
+		)
+	} else {
+		ncv = item
+	}
+	if err := c.handleConsensusMessages(msg, ncv); err != nil {
+		return err
+	}
+	if err := c.handlePostConsensusMessages(msg, ncv); err != nil {
+		return err
+	}
+	return nil
+}
 
-			opts := c.validatorOptions
-			opts.SSVShare = share
-			operator, err := c.committeeMemberFromShare(opts.SSVShare)
-			if err != nil {
-				return err
-			}
-			opts.Operator = operator
+func (c *controller) handleConsensusMessages(msg *queue.DecodedSSVMessage, ncv *committeeObserver) error {
+	c.committeesObserversMutex.Lock()
+	defer c.committeesObserversMutex.Unlock()
+	if msg.MsgType != spectypes.SSVConsensusMsgType {
+		return nil
+	}
 
-			ncv = &nonCommitteeValidator{
-				NonCommitteeValidator: validator.NewNonCommitteeValidator(c.logger, msg.GetID(), opts),
-			}
+	if msg.MsgID.GetRoleType() != spectypes.RoleCommittee {
+		return nil
+	}
 
-			ttlSlots := nonCommitteeValidatorTTLs[msg.MsgID.GetRoleType()]
-			c.nonCommitteeValidators.Set(
-				msg.GetID(),
-				ncv,
-				time.Duration(ttlSlots)*c.beacon.GetBeaconNetwork().SlotDurationSec(),
-			)
-		}
+	subMsg, ok := msg.Body.(*specqbft.Message)
+	if !ok || subMsg.MsgType != specqbft.ProposalMsgType {
+		return nil
+	}
+	ncv.OnProposalMsg(msg)
+	return nil
+}
 
-		ncv.Lock()
+func (c *controller) handlePostConsensusMessages(msg *queue.DecodedSSVMessage, ncv *committeeObserver) error {
+	if msg.MsgType != spectypes.SSVPartialSignatureMsgType {
 		return nil
-	}()
-	if err != nil {
+	}
+
+	c.committeesObserversMutex.Lock()
+	defer c.committeesObserversMutex.Unlock()
+
+	pSigMessages := &spectypes.PartialSignatureMessages{}
+	if err := pSigMessages.Decode(msg.SignedSSVMessage.SSVMessage.GetData()); err != nil {
 		return err
 	}
 
-	// Process the message.
-	defer ncv.Unlock()
-	ncv.ProcessMessage(c.logger, msg)
+	ncv.ProcessMessage(msg)
+	return nil
+}
```

Comment:
You only need one function here, `handleNonCommitteeMessage` you can also lock and unlock the mutex outside of the function.